ci/list-integration-tests-for-modified-modules.sh

```diff
@@ -0,0 +1,50 @@
+#!/bin/bash -e
+# Lists test modules based on git diff check.
+
+if (( $# < 2 ))
+then
+	echo 1>&2 "Usage: $0 <main branch, which we'll compare to to identify changes> <branch with changes we want to identify>"
+	echo 1>&2 "Aborting."
+	exit 1
+fi
+
+# `exec` will echo a based dir and the artifact id.
+#   Basedir is going to be used to figure out to which module folder the modified files belong to
+#   Artifact ID is going to be used to match the basedir to an id that can be later passed as a mvn build parameter
+#
+# The output format is `artifact-id-and-basedir-output :artifactId=basedir`.
+# The prefix (`artifact-id-and-basedir-output`) is used for easier identification of the lines we are interested in from maven's output
+readarray -t lines < <(./mvnw exec:exec@print-artifact-id-and-basedir | grep artifact-id-and-basedir-output | awk '{print $2}')
+
+declare -a artifacts
+declare -a paths
+
+# We split the `:artifactId=basedir` and put them in their own arrays.
+# Basedir will be used as patterns to determine which module the change belongs to.
+for index in "${!lines[@]}"
+do
+	artifacts[index]=${lines[$index]%=*}
+	paths[index]=${lines[$index]#*=}
+	paths[index]="${paths[index]//${paths[0]}"/"/""}"
+done
+
+# get the diffs, which will produce relative paths of modified files:
+readarray -t lines < <(git --no-pager diff --name-only $1..$2)
+
+declare -A modifiedArtifactIds
+
+# using an associated array as a set to have a unique collection of modified artifact IDs
+for modifiedFile in "${lines[@]}"
+do
+  for index in "${!paths[@]}"
+    do
+      if [[ $modifiedFile =~ ${paths[$index]} ]]; then
+          modifiedArtifactIds[${artifacts[$index]}]="modified"
+      fi
+    done
+done
+
+if [ "${#modifiedArtifactIds[@]}" -ne 0 ]; then
+   $(dirname "$(realpath "$0")")/list-dependent-integration-tests.sh "${!modifiedArtifactIds[@]}"
```

Comment:
That would be good for the main build, but not exactly what we want for HSEARCH-4784, I think?

We'd have e.g. an "Elasticsearch 7.17" job that only executes integration tests for anything that depends (even indirectly) on `hibernate-search-backend-elasticsearch`. So we'd actually want to test whether the list of modified artifacts contains `hibernate-search-backend-elasticsearch` or a dependency of `hibernate-search-backend-elasticsearch`... 

Another way of putting it: we'd want to execute integration tests that are listed in **both** `list-dependent-integration-tests.sh "${!modifiedArtifactIds[@]}"` and `list-dependent-integration-tests.sh `:hibernate-search-backend-elasticsearch`. But maybe that's what you intended to do, as a post-processing step on the result of this script?

Reply:
> if we want the list of tests ... then we probably need to `mvn ... -amd -pl :list of modified modules`

I think `list-dependent-integration-tests.sh "${!modifiedArtifactIds[@]}"` like you did is probably fine for our purposes. The only difference is that we wouldn't list non-IT modules, but who cares :)

Looks like this would do what we want:

```
grep -xFf <(ci/list-integration-tests-for-modified-modules.sh $CHANGE_TARGET HEAD) <(list-dependent-integration-tests.sh :hibernate-search-backend-elasticsearch)
```

That's a rather complex command though, maybe it would be better to do this directly in your new script and rename it to something like `list-dependent-integration-tests-for-modified-modules.sh <base> <target> <module> [<other modules> ...]` or something.